### Regression Assignment
In this assignment, we are going to use regression to predict the selling price of a vehicle based on its attributes such as year, mileage, condition. The following features will be considered for our input: odometer, year, condition. These will be our numerical data. We will also look at some categorical data such as make and model. Our target variable will be selling price. We will be utilizing three different regression models on our dataset. They are as follows: linear regression, decision tree regression, and random forest regression. 

In [2]:
import pandas as pd

# Load the data
data = pd.read_csv('/Users/khali/Desktop/Coding Workspace/DataMining/data/cleaned/cleaned_data.csv')  

In [1]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score

# Load the data
data = pd.read_csv('/Users/khali/Desktop/Coding Workspace/DataMining/data/cleaned/cleaned_data.csv')  

# Separate features (X) and target variable (y)
X = data[['odometer', 'year', 'condition', 'make', 'model']]  
y = data['sellingprice']  

# One-Hot Encode categorical columns in X
X = pd.get_dummies(X, columns=['make', 'model'], drop_first=True)  # Modify columns as necessary

# Split data into training and testing sets (80/20 split)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Display the shapes of the resulting splits to confirm
print("Training set shape:", X_train.shape, y_train.shape)
print("Testing set shape:", X_test.shape, y_test.shape)



Training set shape: (377860, 822) (377860,)
Testing set shape: (94465, 822) (94465,)


In [ ]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score

# Initialize the Linear Regression model
linear_model = LinearRegression()

# Train the model
linear_model.fit(X_train, y_train)

# Make predictions
y_train_pred = linear_model.predict(X_train)
y_test_pred = linear_model.predict(X_test)

# Evaluate performance
print("Linear Regression:")
print("Train MSE:", mean_squared_error(y_train, y_train_pred))
print("Test MSE:", mean_squared_error(y_test, y_test_pred))
print("Train MAE:", mean_absolute_error(y_train, y_train_pred))
print("Test MAE:", mean_absolute_error(y_test, y_test_pred))
print("Train R^2:", r2_score(y_train, y_train_pred))
print("Test R^2:", r2_score(y_test, y_test_pred))


Linear Regression:
Train MSE: 17069826.508860268
Test MSE: 16499594.093299313
Train MAE: 2659.2779756416912
Test MAE: 2635.2835306298102
Train R^2: 0.8160382999939508
Test R^2: 0.8184662661238347


In [7]:
from sklearn.tree import DecisionTreeRegressor

# Initialize the Decision Tree Regressor model
tree_model = DecisionTreeRegressor(max_depth=10, random_state=42)  # max_depth can be adjusted

# Train the model
tree_model.fit(X_train, y_train)

# Make predictions
y_train_pred = tree_model.predict(X_train)
y_test_pred = tree_model.predict(X_test)

# Evaluate performance
print("\nDecision Tree Regression:")
print("Train MSE:", mean_squared_error(y_train, y_train_pred))
print("Test MSE:", mean_squared_error(y_test, y_test_pred))
print("Train MAE:", mean_absolute_error(y_train, y_train_pred))
print("Test MAE:", mean_absolute_error(y_test, y_test_pred))
print("Train R^2:", r2_score(y_train, y_train_pred))
print("Test R^2:", r2_score(y_test, y_test_pred))



Decision Tree Regression:
Train MSE: 35254419.82503211
Test MSE: 35836858.96192157
Train MAE: 4020.6561373110503
Test MAE: 4036.826267248772
Train R^2: 0.620062746368657
Test R^2: 0.6057115841175058


In [8]:
from sklearn.ensemble import RandomForestRegressor

# Initialize the Random Forest Regressor model
forest_model = RandomForestRegressor(n_estimators=100, max_depth=10, random_state=42)

# Train the model
forest_model.fit(X_train, y_train)

# Make predictions
y_train_pred = forest_model.predict(X_train)
y_test_pred = forest_model.predict(X_test)

# Evaluate performance
print("\nRandom Forest Regression:")
print("Train MSE:", mean_squared_error(y_train, y_train_pred))
print("Test MSE:", mean_squared_error(y_test, y_test_pred))
print("Train MAE:", mean_absolute_error(y_train, y_train_pred))
print("Test MAE:", mean_absolute_error(y_test, y_test_pred))
print("Train R^2:", r2_score(y_train, y_train_pred))
print("Test R^2:", r2_score(y_test, y_test_pred))



Random Forest Regression:
Train MSE: 33589047.16368291
Test MSE: 33988451.93636097
Train MAE: 3943.298764313782
Test MAE: 3953.1196536843054
Train R^2: 0.6380104850739318
Test R^2: 0.6260483407174287


In [9]:
results = {
    "Linear Regression": {
        "Train MSE": mean_squared_error(y_train, linear_model.predict(X_train)),
        "Test MSE": mean_squared_error(y_test, linear_model.predict(X_test)),
        "Train R^2": r2_score(y_train, linear_model.predict(X_train)),
        "Test R^2": r2_score(y_test, linear_model.predict(X_test)),
    },
    "Decision Tree": {
        "Train MSE": mean_squared_error(y_train, tree_model.predict(X_train)),
        "Test MSE": mean_squared_error(y_test, tree_model.predict(X_test)),
        "Train R^2": r2_score(y_train, tree_model.predict(X_train)),
        "Test R^2": r2_score(y_test, tree_model.predict(X_test)),
    },
    "Random Forest": {
        "Train MSE": mean_squared_error(y_train, forest_model.predict(X_train)),
        "Test MSE": mean_squared_error(y_test, forest_model.predict(X_test)),
        "Train R^2": r2_score(y_train, forest_model.predict(X_train)),
        "Test R^2": r2_score(y_test, forest_model.predict(X_test)),
    }
}

# Print or analyze results
for model, metrics in results.items():
    print(f"\n{model}:")
    for metric, value in metrics.items():
        print(f"{metric}: {value}")



Linear Regression:
Train MSE: 17069826.508860268
Test MSE: 16499594.093299313
Train R^2: 0.8160382999939508
Test R^2: 0.8184662661238347

Decision Tree:
Train MSE: 35254419.82503211
Test MSE: 35836858.96192157
Train R^2: 0.620062746368657
Test R^2: 0.6057115841175058

Random Forest:
Train MSE: 33589047.16368291
Test MSE: 33988451.93636097
Train R^2: 0.6380104850739318
Test R^2: 0.6260483407174287


### Regression Assignment Questions

1) Our input data were the following: odometer, year, condition, make, and model. Our target variable was the selling price.

2) We tried the linear regression, decision tree, and random forest models. For the linear regression, the train model had a R-squared value of 0.816 and the test model had a R-squared value of 0.818. Since these values are really similar, we can conclude that this regression model does a good job getting accurate predictions on untrained data. The decision tree regression model did not perform as well and had R-squared values of 0.638 and 0.626 for the training and testing data respectively. This model also had much higher mean squared error values which showcases worse predictions. The random forest model did slightly better than the decision tree model. It had R-squared values of 0.638 for the trainign data and 0.626 for the tetsing data.

3) Since the R-squared values for the training and testing data were pretty close, this suggests no overfitting or underfitting occured with the regression model. In regards to the decision tree model, the low R-squared values in addition to the really high mean squared error values can indicate some underfitting, which means that no significant patterns are being analyzed from the data. As for the random forest model, it does predict better than the decision tree model, but not as good as the linear regression model. This can indicate slight underfitting, but not overfitting. 

4) We're going to try and modify the decision tree model so it does not perform any underfitting. Here is what we did, we increased the max depth from 10 to 20 in order for the model to find more patterns. The model also can now split at multiple nodes and create branches with only one leaf. This should decrease our mean squared error values and increase our R-squared values. 

In [ ]:
# Question 4 Continued
from sklearn.tree import DecisionTreeRegressor
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score

tree_model_complex = DecisionTreeRegressor(
    max_depth=20,  # Increase the max depth to capture more complex patterns
    min_samples_split=2,  # Allow splits at every node (lowering this increases complexity)
    min_samples_leaf=1,  # Allow smaller leaf nodes (increases complexity)
    random_state=42
)

# Train the model
tree_model_complex.fit(X_train, y_train)

# Make predictions
y_train_pred_complex = tree_model_complex.predict(X_train)
y_test_pred_complex = tree_model_complex.predict(X_test)

# Evaluate performance
print("\nComplex Decision Tree Regression:")
print("Train MSE:", mean_squared_error(y_train, y_train_pred_complex))
print("Test MSE:", mean_squared_error(y_test, y_test_pred_complex))
print("Train MAE:", mean_absolute_error(y_train, y_train_pred_complex))
print("Test MAE:", mean_absolute_error(y_test, y_test_pred_complex))
print("Train R^2:", r2_score(y_train, y_train_pred_complex))
print("Test R^2:", r2_score(y_test, y_test_pred_complex))



Complex Decision Tree Regression:
Train MSE: 20206148.47625138
Test MSE: 23253333.197020967
Train MAE: 2914.5785656696426
Test MAE: 3232.1096069530768
Train R^2: 0.7822381251305388
Test R^2: 0.7441595001396966


### Regression Assignment Questions Continued 

4) After modifying the model, we can see a significant difference in our mean squared error and R-squared values. The MSE values for the training data dropped from 35 million to 20 and the testing data dropped from 35 million to 23 million which indicates a more balanced fit. The R-squared values for the training and testing data also increased to 0.782 and 0.744 respectively. Since these values are fairly close, we can conclude that there is no overfitting occuring here as well as no underfitting due to the increased complexity in the pattern analysis leading to better predictions. 

5) In order to increase the performance of our models from the data perspective, we can do the following: gather more data, clean the data and fine-tune it better, find and determine outliers, try better encoding methods for categorical variables, and maybe incorporate a time-based analysis which can go over the sale date of vehicles to better find trends and patterns. 